filters/auth/oidc.go

```diff
@@ -0,0 +1,565 @@
+	letterBytes   = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
+
+var (
+	src      = rand.NewSource(time.Now().UnixNano())
+	stateMap = make(map[string]bool)
+)
+
+// https://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-golang
+func randString(n int) string {
+	b := make([]byte, n)
+	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!
+	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
+		if remain == 0 {
+			cache, remain = src.Int63(), letterIdxMax
+		}
+		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
+			b[i] = letterBytes[idx]
+			i--
+		}
+		cache >>= letterIdxBits
+		remain--
+	}
+
+	return string(b)
+}
+
+func getCiphersuite() (cipher.AEAD, error) {
+	password := getPassword()
+	salt := getSalt()
+
+	key, err := scrypt.Key(password, salt, 1<<15, 8, 1, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create key: %v", err)
+	}
+	//key has to be 16 or 32 byte
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new cipher: %v", err)
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new GCM: %v", err)
+	}
+	return aesgcm, nil
+}
+
+// getPassword returns a secret byte slice to use as secret encryption key
+// TODO(sszuecs): add enc/dec cipher support and get all keymaterial from trusted sources
+func getPassword() []byte {
+	return []byte("supersecret")
+}
+
+// getSalt returns an 8 byte salt
+// TODO(sszuecs): get salt from trusted sources
+func getSalt() []byte {
+	return []byte{0xc8, 0x28, 0xf2, 0x58, 0xa7, 0x6a, 0xad, 0x7b}
+}
+
+func getTimestampFromState(b []byte, nonceLength int) time.Time {
+	log.Debugf("getTimestampFromState b: %s", b)
+	if len(b) <= secretSize+nonceLength || secretSize >= len(b)-nonceLength {
+		log.Debugf("wrong b: %d, %d, %d, b[%d : %d], %v %v", len(b), secretSize, nonceLength, secretSize, len(b)-nonceLength, len(b) <= secretSize+nonceLength, secretSize >= len(b)-nonceLength)
+		return time.Time{}.Add(1 * time.Second)
+	}
+	ts := string(b[secretSize : len(b)-nonceLength])
+	i, err := strconv.Atoi(ts)
+	if err != nil {
+		log.Errorf("Atoi failed: %v", err)
+		return time.Time{}
+	}
+	return time.Unix(int64(i), 0)
+
+}
+
+func createState(nonce string) string {
+	return randString(secretSize) + fmt.Sprintf("%d", time.Now().Add(1*time.Minute).Unix()) + nonce
+}
+
+func (f *tokenOidcFilter) doRedirect(ctx filters.FilterContext) {
+	nonce, err := f.createNonce()
+	if err != nil {
+		log.Errorf("Failed to create nonce: %v", err)
+		return
+	}
+
+	statePlain := createState(fmt.Sprintf("%x", nonce))
+	stateEnc, err := f.encryptDataBlock([]byte(statePlain))
```

Comment:
the encrypted cookie produced here would provide much better security if you append the caller's IP and user agent (either verify the user agent format, or hash it, to prevent the attacker being able to ask you to sign whatever data they want), then sign it, and include the signature, but not ip or user agent. so you encrypt statePlain+sign(statePlain+ip+hash(userAgent)) where '+' is a concatenate operator.